jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java

```diff
@@ -51,13 +55,80 @@ public Class findClass(@Nullable String name) {
     }
   }
 
+  /**
+   * If {@code mainClass} is {@code null}, tries to infer main class in this order:
+   *
+   * <ul>
+   *   <li>1. Looks in a {@code jar} plugin provided by {@code projectProperties} ({@code
+   *       maven-jar-plugin} for maven or {@code jar} task for gradle).
+   *   <li>2. Searches for a class defined with a main method.
+   * </ul>
+   *
+   * <p>Warns if main class is not valid, or throws an error if no valid main class is not found.
+   */
+  public static String resolveMainClass(
+      @Nullable String mainClass, ProjectProperties projectProperties) {
+    BuildLogger logger = projectProperties.getLogger();
+    if (mainClass == null) {
+      logger.info(
+          "Searching for main class... Add a 'mainClass' configuration to '"
+              + projectProperties.getPluginName()
+              + "' to improve build speed.");
+      mainClass = projectProperties.getMainClassFromJar();
+      if (mainClass == null || !BuildConfiguration.isValidJavaClass(mainClass)) {
+        logger.debug(
+            "Could not find a valid main class specified in a jar plugin; attempting to infer "
+                + "main class.");
+
+        try {
+          // Adds each file in each classes output directory to the classes files list.
+          List<String> mainClasses = new ArrayList<>();
+          for (Path classPath : projectProperties.getSourceFilesConfiguration().getClassesFiles()) {
+            mainClasses.addAll(findMainClasses(classPath));
+          }
+
+          if (mainClasses.size() == 1) {
+            // Valid class found; use inferred main class
+            mainClass = mainClasses.get(0);
+          } else if (mainClasses.size() == 0 && mainClass == null) {
+            // No main class found anywhere
+            throw new IllegalStateException(
```

Comment:
Oh I just noticed this should prob be wrapped in the `GradleException`/`MojoExecutionException` (wrap-unwrap-rewrap a thrown checked exception)